From aebec11e15067ea59ebf34dbf88166cb17724157 Mon Sep 17 00:00:00 2001
From: mancha <mancha1 AT zoho DOT com>
Date: Sat, 9 May 2015
Subject: Fix issues with detail::copy_file

For compiled operational function detail::copy_file, pass
detail::copy_options, a plain-old enum, rather than a BOOST_SCOPED_ENUM.
We cannot pass a BOOST_SCOPED_ENUM to a compled function because it will
result in an undefined reference if the library is compiled with
-std=c++0x but the use is compiled in C++03 mode, or vise versa. 

Backport of upstream fix for use with Boost 1.56.0.

---
 boost/filesystem/operations.hpp    |   31 ++++++++++++++++++-------
 libs/filesystem/src/operations.cpp |    8 ++----
 2 files changed, 25 insertions(+), 14 deletions(-)

--- a/boost/filesystem/operations.hpp
+++ b/boost/filesystem/operations.hpp
@@ -302,7 +302,7 @@ namespace boost
   };
 
   BOOST_SCOPED_ENUM_START(copy_option)
-    {none, fail_if_exists = none, overwrite_if_exists};
+    {none=0, fail_if_exists = none, overwrite_if_exists};
   BOOST_SCOPED_ENUM_END
 
 //--------------------------------------------------------------------------------------//
@@ -311,6 +311,11 @@ namespace boost
 
   namespace detail
   {
+    //  We cannot pass a BOOST_SCOPED_ENUM to a compled function because it will result
+    //  in an undefined reference if the library is compled with -std=c++0x but the use
+    //  is compiled in C++03 mode, or visa versa. See tickets 6124, 6779, 10038.
+    enum copy_option {none=0, fail_if_exists = none, overwrite_if_exists};
+
     BOOST_FILESYSTEM_DECL
     file_status status(const path&p, system::error_code* ec=0);
     BOOST_FILESYSTEM_DECL
@@ -326,9 +331,8 @@ namespace boost
     BOOST_FILESYSTEM_DECL
     void copy_directory(const path& from, const path& to, system::error_code* ec=0);
     BOOST_FILESYSTEM_DECL
-    void copy_file(const path& from, const path& to,
-                    BOOST_SCOPED_ENUM(copy_option) option,  // See ticket #2925
-                    system::error_code* ec=0);
+    void copy_file(const path& from, const path& to,  // See ticket #2925
+                   detail::copy_option option, system::error_code* ec=0);
     BOOST_FILESYSTEM_DECL
     void copy_symlink(const path& existing_symlink, const path& new_symlink, system::error_code* ec=0);
     BOOST_FILESYSTEM_DECL
@@ -489,19 +493,28 @@ namespace boost
   inline
   void copy_file(const path& from, const path& to,   // See ticket #2925
                  BOOST_SCOPED_ENUM(copy_option) option)
-                                       {detail::copy_file(from, to, option);}
+  {
+    detail::copy_file(from, to, static_cast<detail::copy_option>(option));
+  }
   inline
   void copy_file(const path& from, const path& to)
-                                       {detail::copy_file(from, to, copy_option::fail_if_exists);}
+  {
+    detail::copy_file(from, to, detail::fail_if_exists);
+  }
   inline
   void copy_file(const path& from, const path& to,   // See ticket #2925
                  BOOST_SCOPED_ENUM(copy_option) option, system::error_code& ec)
-                                       {detail::copy_file(from, to, option, &ec);}
+  {
+    detail::copy_file(from, to, static_cast<detail::copy_option>(option), &ec);
+  }
   inline
   void copy_file(const path& from, const path& to, system::error_code& ec)
-                                       {detail::copy_file(from, to, copy_option::fail_if_exists, &ec);}
+  {
+    detail::copy_file(from, to, detail::fail_if_exists, &ec);
+  }
   inline
-  void copy_symlink(const path& existing_symlink, const path& new_symlink) {detail::copy_symlink(existing_symlink, new_symlink);}
+  void copy_symlink(const path& existing_symlink,
+                    const path& new_symlink) {detail::copy_symlink(existing_symlink, new_symlink);}
 
   inline
   void copy_symlink(const path& existing_symlink, const path& new_symlink, system::error_code& ec)
--- a/libs/filesystem/src/operations.cpp
+++ b/libs/filesystem/src/operations.cpp
@@ -869,7 +869,7 @@ namespace detail
     }
     else if(is_regular_file(s))
     {
-      copy_file(from, to, copy_option::fail_if_exists, *ec);
+      copy_file(from, to, fs::copy_option::fail_if_exists, *ec);
     }
     else
     {
@@ -891,12 +891,10 @@ namespace detail
   }
 
   BOOST_FILESYSTEM_DECL
-  void copy_file(const path& from, const path& to,
-                  BOOST_SCOPED_ENUM(copy_option)option,
-                  error_code* ec)
+  void copy_file(const path& from, const path& to, copy_option option, error_code* ec)
   {
     error(!BOOST_COPY_FILE(from.c_str(), to.c_str(),
-      option == copy_option::fail_if_exists),
+      option == fail_if_exists),
         from, to, ec, "boost::filesystem::copy_file");
   }
 
